File: Proxies/Proxies.Retail.TypeScript/Managers/ManagerFactory.ts
/**
 * SAMPLE CODE NOTICE
 * 
 * THIS SAMPLE CODE IS MADE AVAILABLE AS IS.  MICROSOFT MAKES NO WARRANTIES, WHETHER EXPRESS OR IMPLIED,
 * OF FITNESS FOR A PARTICULAR PURPOSE, OF ACCURACY OR COMPLETENESS OF RESPONSES, OF RESULTS, OR CONDITIONS OF MERCHANTABILITY.
 * THE ENTIRE RISK OF THE USE OR THE RESULTS FROM THE USE OF THIS SAMPLE CODE REMAINS WITH THE USER.
 * NO TECHNICAL SUPPORT IS PROVIDED.  YOU MAY NOT DISTRIBUTE THIS CODE UNLESS YOU HAVE A LICENSE AGREEMENT WITH MICROSOFT THAT ALLOWS YOU TO DO SO.
 */

///<reference path='../DataJS.d.ts'/>
///<reference path='../Tracer.ts'/>
///<reference path='../CommerceContext.g.ts'/>
///<reference path='../Interfaces/IDataServiceRequestFactory.ts'/>
///<reference path='../RetailServerRequestFactory.ts'/>
///<reference path='../XhrHelper.ts'/>

module Commerce.Proxy {
    "use strict";

    export class ManagerFactory implements Commerce.Proxy.IManagerFactory {

        private _commerceContext: Commerce.Proxy.CommerceContext;
        private _odataEndpoint: string;
        private _operatingUnitNumber: string;
        private _authToken: string;
        private _locale: string;

        constructor(retailServerUrl: string, operatingUnitNumber?: string, authToken?: string, locale: string = "") {

            this._odataEndpoint = StringExtensions.CleanUri(retailServerUrl);
            this._operatingUnitNumber = operatingUnitNumber;
            this._authToken = authToken;
            this._locale = locale;

            XhrHelper.SetupAjaxParameters();
            XhrHelper.SetupODataParameters();

            this._commerceContext = new Commerce.Proxy.CommerceContext(this.createDataServiceFactory());
        }

        /**
         * Updates the server Uri of the commerce context.
         *
         * @param {string} serverUri The new URI.
         */
        updateServerUriInCommerceContext(retailServerUri: string) {
            this._odataEndpoint = StringExtensions.CleanUri(retailServerUri);
            this._commerceContext.dataServiceRequestFactory = this.createDataServiceFactory();
        }

        /**
         * Updates the locale of the commerce context.
         * @param {string} locale The new locale.
         */
        updateLocaleInCommerceContext(locale: string): void {
            this._locale = locale;
            this._commerceContext.dataServiceRequestFactory = this.createDataServiceFactory();
        }

        /**
         * Creates an instance of given entity manager.
         *
         * @param {string} entityManagerInterface The interface name.
         */
        public getManager<T>(entityManagerInterface: string): T {
            return <T>this.GetManager(entityManagerInterface);
        }

        /**
         * Creates an instance of given entity manager.
         *
         * @param {string} entityManagerInterface The interface name.
         * @param {any} [callerContext] The optional reference to caller object.
         */
        public GetManager(entityManagerInterface: string, callerContext?: any): any {
            var dataManager: any;

            switch (entityManagerInterface) {

                case Commerce.Proxy.ICartManagerName:
                    dataManager = new Commerce.Proxy.CartManager(this._commerceContext, callerContext);
                    break;

                case Commerce.Proxy.ICustomerManagerName:
                    dataManager = new Commerce.Proxy.CustomerManager(this._commerceContext, callerContext);
                    break;

                case Commerce.Proxy.IProductManagerName:
                    dataManager = new Commerce.Proxy.ProductManager(this._commerceContext, callerContext);
                    break;

                case Commerce.Proxy.ISalesOrderManagerName:
                    dataManager = new Commerce.Proxy.SalesOrderManager(this._commerceContext, callerContext);
                    break;

                case Commerce.Proxy.IStockCountJournalManagerName:
                    dataManager = new Commerce.Proxy.StockCountJournalManager(this._commerceContext, callerContext);
                    break;

                case Commerce.Proxy.IStoreOperationsManagerName:
                    dataManager = new Commerce.Proxy.StoreOperationsManager(this._commerceContext, callerContext);
                    break;

                case Commerce.Proxy.IOrgUnitManagerName:
                    dataManager = new Commerce.Proxy.OrgUnitManager(this._commerceContext, callerContext);
                    break;

                default:
                    throw entityManagerInterface + " is not supported.";
            }

            return dataManager;
        }

        private createDataServiceFactory(): Common.IDataServiceRequestFactory {
            var onlineConfigurationName = "online";
            var offlineConfigurationName = "offline";
            var dataServiceRequestFactory: Common.IDataServiceRequestFactory;

            dataServiceRequestFactory = new Commerce.Proxy.RetailServerRequestFactory(this._odataEndpoint, this._operatingUnitNumber, this._authToken);
            dataServiceRequestFactory.locale = this._locale;

            return dataServiceRequestFactory;
        }
    }
}

File: Proxies/RetailProxy/ManagerFactory.cs
/**
 * SAMPLE CODE NOTICE
 * 
 * THIS SAMPLE CODE IS MADE AVAILABLE AS IS.  MICROSOFT MAKES NO WARRANTIES, WHETHER EXPRESS OR IMPLIED,
 * OF FITNESS FOR A PARTICULAR PURPOSE, OF ACCURACY OR COMPLETENESS OF RESPONSES, OF RESULTS, OR CONDITIONS OF MERCHANTABILITY.
 * THE ENTIRE RISK OF THE USE OR THE RESULTS FROM THE USE OF THIS SAMPLE CODE REMAINS WITH THE USER.
 * NO TECHNICAL SUPPORT IS PROVIDED.  YOU MAY NOT DISTRIBUTE THIS CODE UNLESS YOU HAVE A LICENSE AGREEMENT WITH MICROSOFT THAT ALLOWS YOU TO DO SO.
 */

namespace Contoso
{
    namespace Commerce.RetailProxy
    {
        using System;
        using System.Collections.Generic;
        using System.Diagnostics.CodeAnalysis;
        using System.Linq;
        using System.Reflection;
        using Commerce.RetailProxy.Authentication;
    
        /// <summary>
        /// Class encapsulates manager factory.
        /// </summary>
        public sealed class ManagerFactory
        {
            private static IEnumerable<Type> managerTypes;
    
            /// <summary>
            /// Initializes static members of the <see cref="ManagerFactory"/> class.
            /// </summary>
            [SuppressMessage("Microsoft.Performance", "CA1810:InitializeReferenceTypeStaticFieldsInline", Justification = "Better for code structure.")]
            static ManagerFactory()
            {
                var entityManagerType = typeof(IEntityManager);
    
                // Gets all manager types defined in the assembly.
                managerTypes = from t in typeof(ManagerFactory).GetTypeInfo().Assembly.DefinedTypes
                               where t.Namespace == typeof(ManagerFactory).Namespace
                                    && t.IsClass
                                    && t.ImplementedInterfaces.Any(i => entityManagerType.GetTypeInfo().IsAssignableFrom(i.GetTypeInfo()))
                               select t.AsType();
            }
    
            /// <summary>
            /// Initializes a new instance of the <see cref="ManagerFactory"/> class.
            /// </summary>
            /// <param name="context">The context.</param>
            private ManagerFactory(IContext context)
            {
                this.Context = context;
            }
    
            /// <summary>
            /// Gets the context.
            /// </summary>
            /// <value>
            /// The context.
            /// </value>
            public IContext Context { get; private set; }
    
            /// <summary>
            /// Creates an instance of commerce client with specified configuration.
            /// </summary>
            /// <param name="context">The context.</param>
            /// <returns>The instance of commerce client.</returns>
            public static ManagerFactory Create(IContext context)
            {
                return new ManagerFactory(context);
            }
    
            /// <summary>
            /// Gets the manager.
            /// </summary>
            /// <typeparam name="T">The type of entity manager.</typeparam>
            /// <returns>The instance of the manager.</returns>
            public T GetManager<T>() where T : IEntityManager
            {
                var requiredManagerTypeInfo = typeof(T).GetTypeInfo();
                var managerType = managerTypes.Where(m => requiredManagerTypeInfo.IsAssignableFrom(m.GetTypeInfo())).FirstOrDefault();
    
                if (managerType != null)
                {
                    return (T)Activator.CreateInstance(managerType, this.Context);
                }
    
                throw new NotImplementedException(string.Format("'{0}' is not implemented.", requiredManagerTypeInfo.Name));
            }
        }
    }
}
